keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,832 @@
+```
 
-# gRPC Definitions
+###### 2. User creates BucketAccess to generate credentials for that bucket
 
-There is one service defined by COSI - `Provisioner`. This will need to be satisfied by the vendor-provisioner in order to be COSI-compatible
+Unlike the BucketAccess for COSI created bucket, this BucketClaim should directly reference the Bucket instead of a BucketClaim
 
 ```
-service Provisioner {
-    rpc ProvisionerGetInfo (ProvisionerGetInfoRequest) returns (ProvisionerGetInfoResponse) {}
+    BucketAccess - bac2
+    |-------------------------------|
+    | spec:                         |
+    |   bucketName: bucketName123   |
+    |                               |
+    |-------------------------------|
+```
+This BucketClaim can be specified in pod definition as described in the above sections.
+
+Note that, as of the alpha version of COSI, there is no authorization mechanism to restrict users who can refer to buckets created by other namesapces. In the future, we are working on a design based on `referencePolicy`, a new development in K8s, meant to be a general approach to sharing resources across namespaces.
+
+# Usability
+
+## Self Service
+
+Self service is easily possible with the current design as both the BucketRequest and BucketClaim resources are namespace scoped, and users need not have admin privileges to create, modify and delete them.
+
+The only admin step is the creation of BucketClass and BucketAccessClass. This requirement is no different from requiring a StorageClass for provisioning PVCs. It is a well-understood pattern among kubernetes users.
+
+## Mutating Buckets
+
+As of the current design of COSI, mutating bucket properties is not supported. However, the current design does not prevent us from supporting it in the future. Mutable properties include encryption, object lifecycle configuration, replication configuration etc. These properties will be supported in future versions along with the capability to mutate them.
+
+These properties will be specified in the BucketRequest and follow the same pattern of events as Bucket creation. i.e. Bucket API object will be updated to reflect the properties set in BucketRequest, and then a controller will pick-up these changes and call the appropriate APIs to reflect them in the backend.
+
+# Object Lifecycle
+
+The following resources are managed by admins
+
+- BucketClass
+- BucketAccessClass
+
+The following resources are managed by a user. These are created with a reference to their corresponding class objects:
+
+- BucketClaim -> BucketClass
+- BucketAccess -> Bucket, BucketClaim, BucketAccessClass
+
+The COSI controller responds by creating the following objects
+
+- BucketRequest <-> Bucket (1:1)
+- BucketClaim -> Bucket
+
+Notes:
+
+ - There are **NO** cycles in the relationship graph of the above mentioned API objects.
+ - Mutations are not supported in the API.
+
+When a user deletes the BucketRequest, then depending on the DeletionPolicy, the following happens:
```

Comment:
Has there been any thought to how a retention period for compliance use cases can be supported? Will that conflict with deletion policy?

Reply:
Policies like retention, object locking etc. can be specified using params field. This does not conflict with deletion policy, since deletion policy is orchestrated entirely by kubernetes (COSI), and bucket policies like rentention, object locking etc. are orchestrated entirely by the OSP.